TestRegistryClientImageV1/TestRegistryClientImageV2 failure

```
--- FAIL: TestRegistryClientImageV1 (1.46s)
	dockerregistryclient_test.go:222: the default tag "sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19" has not been set on repository "openshift/origin"
FAIL
```
Same error for `TestRegistryClientImageV2`
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/1423/console

I can tell you more:
```
--- FAIL: TestRegistryClientImageV2 (5.27s)
	dockerregistryclient_test.go:222: the default tag "sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19" has not been set on repository "openshift/origin"
```

`TestRegistryClientImageV2` flakes too.

full log:
```
=== RUN   TestRegistryClientImageV2
I0602 11:54:49.069659   25787 client.go:320] Found registry v2 API at https://registry-1.docker.io/v2/
I0602 11:54:52.695341   25787 client.go:555] passed valid auth token, but unable to find tagged image at "https://registry-1.docker.io/v2/openshift/origin-not-found/manifests/latest", 401 map[Content-Type:[application/json; charset=utf-8] Docker-Distribution-Api-Version:[registry/2.0] Www-Authenticate:[Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="repository:openshift/origin-not-found:pull",error="insufficient_scope"] Date:[Thu, 02 Jun 2016 09:54:52 GMT] Content-Length:[158] Strict-Transport-Security:[max-age=31536000]]: {"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"repository","Name":"openshift/origin-not-found","Action":"pull"}]}]}
I0602 11:54:53.796195   25787 client.go:567] unable to find tagged image at "https://registry-1.docker.io/v2/openshift/origin/manifests/sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19", 404 map[Content-Type:[application/json; charset=utf-8] Docker-Distribution-Api-Version:[registry/2.0] Date:[Thu, 02 Jun 2016 09:54:53 GMT] Content-Length:[192] Strict-Transport-Security:[max-age=31536000]]: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown","detail":{"Name":"openshift/origin","Revision":"sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19"}}]}
--- FAIL: TestRegistryClientImageV2 (5.27s)
	dockerregistryclient_test.go:222: the default tag "sha256:52776e94e88f0335b7165113ea40ebdf0c919eb208d357287537544987d14b19" has not been set on repository "openshift/origin"
FAIL
```

and more:
```
=== RUN   TestImageStreamImportDockerHub
I0602 11:50:28.690466   23916 importer.go:305] importing remote Docker repository registry=https://registry-1.docker.io repository=library/redis insecure=false
I0602 11:50:29.232394   23916 credentials.go:30] asked to provide RefreshToken for https://auth.docker.io/token
I0602 11:50:29.232411   23916 credentials.go:25] asked to provide Basic credentials for https://auth.docker.io/token
I0602 11:50:30.508498   23916 importer.go:305] importing remote Docker repository registry=https://registry-1.docker.io repository=library/mysql insecure=false
I0602 11:50:30.509669   23916 credentials.go:30] asked to provide RefreshToken for https://auth.docker.io/token
I0602 11:50:30.509680   23916 credentials.go:25] asked to provide Basic credentials for https://auth.docker.io/token
I0602 11:50:31.010503   23916 importer.go:305] importing remote Docker repository registry=https://registry-1.docker.io repository=mysql/doesnotexistinanyform insecure=false
I0602 11:50:31.011715   23916 credentials.go:30] asked to provide RefreshToken for https://auth.docker.io/token
I0602 11:50:31.011736   23916 credentials.go:25] asked to provide Basic credentials for https://auth.docker.io/token
I0602 11:50:31.269478   23916 client.go:326] Retrying request to a v2 Docker registry after encountering error (1 attempts remaining): unauthorized: authentication required
I0602 11:50:32.400343   23916 client.go:326] Retrying request to a v2 Docker registry after encountering error (0 attempts remaining): unauthorized: authentication required
I0602 11:50:32.531135   23916 importer.go:437] unable to get tag "latest" for repository &importer.importRepository{Ref:api.DockerImageReference{Registry:"", Namespace:"mysql", Name:"doesnotexistinanyform", Tag:"", ID:""}, Registry:(*url.URL)(0xc820348900), Name:"mysql/doesnotexistinanyform", Insecure:false, Tags:[]importer.importTag{importer.importTag{Name:"latest", Image:(*api.Image)(nil), Err:error(nil)}}, Digests:[]importer.importDigest(nil), MaximumTags:0, AdditionalTags:[]string(nil), Err:error(nil)}: errcode.Error{Code:1002, Message:"authentication required", Detail:[]uint8{}}
I0602 11:50:32.531356   23916 importer.go:305] importing remote Docker repository registry=https://registry-1.docker.io repository=library/mongo insecure=false
I0602 11:50:32.535871   23916 credentials.go:30] asked to provide RefreshToken for https://auth.docker.io/token
I0602 11:50:32.535915   23916 credentials.go:25] asked to provide Basic credentials for https://auth.docker.io/token
--- FAIL: TestImageStreamImportDockerHub (5.99s)
	imageimporter_test.go:696: unexpected object: api.ImageImportStatus{Tag:"latest", Status:unversioned.Status{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ListMeta:unversioned.ListMeta{SelfLink:"", ResourceVersion:""}, Status:"Failure", Message:"Internal error occurred: unauthorized: authentication required", Reason:"InternalError", Details:(*unversioned.StatusDetails)(0xc82072eb40), Code:500}, Image:(*api.Image)(nil)}
FAIL
```

--- a/pkg/dockerregistry/client.go
+++ b/pkg/dockerregistry/client.go
@@ -18,6 +18,9 @@ import (
 	"k8s.io/kubernetes/pkg/client/transport"
 	knet "k8s.io/kubernetes/pkg/util/net"
 
+	"github.com/docker/distribution/manifest/schema1"
+	"github.com/docker/distribution/manifest/schema2"
+
 	imageapi "github.com/openshift/origin/pkg/image/api"
 )
 
@@ -471,6 +474,8 @@ func (repo *v2repository) getTags(c *connection) (map[string]string, error) {
 	if err != nil {
 		return nil, fmt.Errorf("error creating request: %v", err)
 	}
+	addAcceptHeader(req)
+
 	if len(repo.token) > 0 {
 		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", repo.token))
 	}
@@ -528,6 +533,7 @@ func (repo *v2repository) getTaggedImage(c *connection, tag, userTag string) (*I
 	if err != nil {
 		return nil, fmt.Errorf("error creating request: %v", err)
 	}
+	addAcceptHeader(req)
 
 	if len(repo.token) > 0 {
 		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", repo.token))
@@ -579,7 +585,7 @@ func (repo *v2repository) getTaggedImage(c *connection, tag, userTag string) (*I
 	if err != nil {
 		return nil, fmt.Errorf("can't read image body from %s: %v", req.URL, err)
 	}
-	dockerImage, err := unmarshalV2DockerImage(body)
+	dockerImage, err := repo.unmarshalImageManifest(c, body)
 	if err != nil {
 		return nil, err
 	}
@@ -595,6 +601,87 @@ func (repo *v2repository) getTaggedImage(c *connection, tag, userTag string) (*I
 
 func (repo *v2repository) getImage(c *connection, image, userTag string) (*Image, error) {
 	return repo.getTaggedImage(c, image, userTag)
+}
+
+func (repo *v2repository) getImageConfig(c *connection, dgst string) ([]byte, error) {
+	endpoint := repo.endpoint
+	endpoint.Path = path.Join(endpoint.Path, fmt.Sprintf("/v2/%s/blobs/%s", repo.name, dgst))
+	req, err := http.NewRequest("GET", endpoint.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating request: %v", err)
+	}
+
+	if len(repo.token) > 0 {
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", repo.token))
+	}
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, convertConnectionError(c.url.String(), fmt.Errorf("error getting image config for %s: %v", repo.name, err))
+	}
+	defer resp.Body.Close()
+
+	switch code := resp.StatusCode; {
+	case code == http.StatusUnauthorized:
+		if len(repo.token) != 0 {
+			// The DockerHub returns JWT tokens that take effect at "now" at second resolution, which means clients can
+			// be rejected when requests are made near the time boundary.
+			if repo.retries > 0 {
+				repo.retries--
+				time.Sleep(time.Second / 2)
+				return repo.getImageConfig(c, dgst)
+			}
+			delete(c.cached, repo.name)
+			// docker will not return a NotFound on any repository URL - for backwards compatibility, return NotFound on the
+			// repo
+			body, _ := ioutil.ReadAll(resp.Body)
+			glog.V(4).Infof("passed valid auth token, but unable to find image config at %q, %d %v: %s", req.URL.String(), resp.StatusCode, resp.Header, body)
+			return nil, errBlobNotFound{dgst, repo.name}
+		}
+		token, err := c.authenticateV2(resp.Header.Get("WWW-Authenticate"))
+		if err != nil {
+			return nil, fmt.Errorf("error getting image config for %s:%s: %v", repo.name, dgst, err)
+		}
+		repo.retries = 2
+		repo.token = token
+		return repo.getImageConfig(c, dgst)
+	case code == http.StatusNotFound:
+		body, _ := ioutil.ReadAll(resp.Body)
+		glog.V(4).Infof("unable to find image config at %q, %d %v: %s", req.URL.String(), resp.StatusCode, resp.Header, body)
+		return nil, errBlobNotFound{dgst, repo.name}
+	case code >= 300 || resp.StatusCode < 200:
+		// token might have expired - evict repo from cache so we can get a new one on retry
+		delete(c.cached, repo.name)
+
+		return nil, fmt.Errorf("error retrieving image config: server returned %d", resp.StatusCode)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("can't read image body from %s: %v", req.URL, err)
+	}
+
+	return body, nil
+}
+
+func (repo *v2repository) unmarshalImageManifest(c *connection, body []byte) (*docker.Image, error) {
+	manifest := imageapi.DockerImageManifest{}
+	if err := json.Unmarshal(body, &manifest); err != nil {
+		return nil, err
+	}
+	switch manifest.SchemaVersion {
+	case 1:
+		if len(manifest.History) == 0 {
+			return nil, fmt.Errorf("image has no v1Compatibility history and cannot be used")
+		}
+		return unmarshalDockerImage([]byte(manifest.History[0].DockerV1Compatibility))
+	case 2:
+		config, err := repo.getImageConfig(c, manifest.Config.Digest)
+		if err != nil {
+			return nil, err
+		}
+		return unmarshalDockerImage(config)
+	}
+	return nil, fmt.Errorf("unrecognized Docker image manifest schema %d", manifest.SchemaVersion)
 }
 
 // v1repository exposes methods for accessing a named Docker V1 repository on a server.
@@ -713,6 +800,16 @@ func (repo *v1repository) getImage(c *connection, image, userTag string) (*Image
 	return &Image{Image: *dockerImage}, nil
 }
 
+// errBlobNotFound is an error indicating the requested blob does not exist in the repository.
+type errBlobNotFound struct {
+	digest     string
+	repository string
+}
+
+func (e errBlobNotFound) Error() string {
+	return fmt.Sprintf("blob %s was not found in repository %q", e.digest, e.repository)
+}
+
 // errTagNotFound is an error indicating the requested tag does not exist on the server. May be returned on
 // a v2 repository when the repository does not exist (because the v2 registry returns 401 on any repository
 // you do not have permission to see, or does not exist)
@@ -784,8 +881,13 @@ func IsTagNotFound(err error) bool {
 	return ok
 }
 
+func IsBlobNotFound(err error) bool {
+	_, ok := err.(errBlobNotFound)
+	return ok
+}
+
 func IsNotFound(err error) bool {
-	return IsRegistryNotFound(err) || IsRepositoryNotFound(err) || IsImageNotFound(err) || IsTagNotFound(err)
+	return IsRegistryNotFound(err) || IsRepositoryNotFound(err) || IsImageNotFound(err) || IsTagNotFound(err) || IsBlobNotFound(err)
 }
 
 func unmarshalDockerImage(body []byte) (*docker.Image, error) {
@@ -809,13 +911,7 @@ func unmarshalDockerImage(body []byte) (*docker.Image, error) {
 	}, nil
 }
 
-func unmarshalV2DockerImage(body []byte) (*docker.Image, error) {
-	manifest := imageapi.DockerImageManifest{}
-	if err := json.Unmarshal(body, &manifest); err != nil {
-		return nil, err
-	}
-	if len(manifest.History) == 0 {
-		return nil, fmt.Errorf("image has no v1Compatibility history and cannot be used")
-	}
-	return unmarshalDockerImage([]byte(manifest.History[0].DockerV1Compatibility))
+func addAcceptHeader(r *http.Request) {
+	r.Header.Add("Accept", schema1.MediaTypeManifest)
+	r.Header.Add("Accept", schema2.MediaTypeManifest)
 }

--- a/test/integration/dockerregistryclient_test.go
+++ b/test/integration/dockerregistryclient_test.go
@@ -182,7 +182,7 @@ func TestRegistryClientRegistryNotFound(t *testing.T) {
 	}
 }
 
-func doTestRegistryClientImage(t *testing.T, registry, version string) {
+func doTestRegistryClientImage(t *testing.T, registry, reponame, version string) {
 	conn, err := dockerregistry.NewClient(10*time.Second, version == "v2").Connect(registry, false)
 	if err != nil {
 		t.Fatal(err)
@@ -198,22 +198,28 @@ func doTestRegistryClientImage(t *testing.T, registry, version string) {
 
 	var image *dockerregistry.Image
 	err = retryWhenUnreachable(t, func() error {
-		image, err = conn.ImageByTag("openshift", "origin", "latest")
+		image, err = conn.ImageByTag("openshift", reponame, "latest")
 		return err
 	})
 	if err != nil {
 		t.Fatal(err)
 	}
-	if len(image.ContainerConfig.Entrypoint) == 0 {
-		t.Errorf("%s: unexpected image: %#v", version, image)
+
+	if image.Comment != "Imported from -" {
+		t.Errorf("%s: unexpected image comment", version)
 	}
+
+	if image.Architecture != "amd64" {
+		t.Errorf("%s: unexpected image architecture", version)
+	}
+
 	if version == "v2" && !image.PullByID {
 		t.Errorf("%s: should be able to pull by ID %s", version, image.ID)
 	}
 
 	var other *dockerregistry.Image
 	err = retryWhenUnreachable(t, func() error {
-		other, err = conn.ImageByID("openshift", "origin", image.ID)
+		other, err = conn.ImageByID("openshift", reponame, image.ID)
 		return err
 	})
 	if err != nil {
@@ -224,14 +230,19 @@ func doTestRegistryClientImage(t *testing.T, registry, version string) {
 	}
 }
 
-func TestRegistryClientImageV2(t *testing.T) {
-	t.Skip("openshift/origin is being pushed by Docker 1.9, which causes this test to fail because digests don't match")
-	doTestRegistryClientImage(t, dockerHubV2RegistryName, "v2")
+func TestRegistryClientAPIv2ManifestV2Schema2(t *testing.T) {
+	t.Log("openshift/schema-v2-test-repo was pushed by Docker 1.11.1")
+	doTestRegistryClientImage(t, dockerHubV2RegistryName, "schema-v2-test-repo", "v2")
 }
 
-func TestRegistryClientImageV1(t *testing.T) {
-	t.Skip("openshift/origin is being pushed by Docker 1.9, which causes this test to fail because digests don't match")
-	doTestRegistryClientImage(t, dockerHubV1RegistryName, "v1")
+func TestRegistryClientAPIv2ManifestV2Schema1(t *testing.T) {
+	t.Log("openshift/schema-v1-test-repo was pushed by Docker 1.8.2")
+	doTestRegistryClientImage(t, dockerHubV2RegistryName, "schema-v1-test-repo", "v2")
+}
+
+func TestRegistryClientAPIv1(t *testing.T) {
+	t.Log("openshift/schema-v1-test-repo was pushed by Docker 1.8.2")
+	doTestRegistryClientImage(t, dockerHubV1RegistryName, "schema-v1-test-repo", "v1")
 }
 
 func TestRegistryClientQuayIOImage(t *testing.T) {

